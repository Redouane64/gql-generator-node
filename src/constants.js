/* DOCUMENT INFORMATION
	- Author:   Dominik Maszczyk
	- Email:    Skitionek@gmail.com
	- Created:  2019-06-06
*/

export const QUERY_KINDS = [
	'Mutation', 'Query', 'Subscription'
];

export const MODULE_NAME = 'gql-generator-node';