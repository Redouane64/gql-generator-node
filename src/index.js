#!/usr/bin/env node

/**
 * Compile arguments dictionary for a field
 * @param field current field object
 * @param duplicateArgCounts map for deduping argument name collisions
 * @param allArgsDict dictionary of all arguments
 */
const getFieldArgsDict = (field,
                          duplicateArgCounts,
                          allArgsDict = {}, path) =>
	field.args.reduce((o, arg) => {
		const arg_name = `${path.join('_')}_${field.name}_${arg.name}`;
		if (arg_name in duplicateArgCounts) {
			const index = duplicateArgCounts[arg_name] + 1;
			duplicateArgCounts[arg_name] = index;
			o[`${arg_name}${index}`] = arg;
		} else if (allArgsDict[arg_name]) {
			duplicateArgCounts[arg_name] = 1;
			o[arg_name] = arg;
		} else if (!path.length) {
			o[arg.name] = arg;
		} else {
			o[arg_name] = arg;
		}
		return o;
	}, {});

module.exports = function (schema, depthLimit = 100, dedupe = getFieldArgsDict) {

	const result = {};

	/**
	 * Generate variables string
	 * @param dict dictionary of arguments
	 */
	const getArgsToVarsStr = dict => Object.entries(dict)
		.map(([varName, arg]) => `${arg.name}: $${varName}`)
		.join(', ');

	/**
	 * Generate types string
	 * @param dict dictionary of arguments
	 */
	const getVarsToTypesStr = dict => Object.entries(dict)
		.map(([varName, arg]) => `$${varName}: ${arg.type}`)
		.join(', ');

	/**
	 * Generate the query for the specified field
	 * @param curName name of the current field
	 * @param curParentType parent type of the current field
	 * @param curParentName parent name of the current field
	 * @param argumentsDict dictionary of arguments from all fields
	 * @param duplicateArgCounts map for deduping argument name collisions
	 * @param crossReferenceKeyList list of the cross reference
	 * @param curDepth currentl depth of field
	 */
	const generateQuery = (curName,
	                       curParentType,
	                       curParentName,
	                       argumentsDict = {},
	                       duplicateArgCounts = {},
	                       crossReferenceKeyList = [], // [`${curParentName}To${curName}Key`]
	                       curDepth = 1,
	                       path = []) => {
		const field = schema.getType(curParentType).getFields()[curName];
		const curTypeName = field.type.inspect().replace(/[[\]!]/g, '');
		const curType = schema.getType(curTypeName);
		let queryStr = '';
		let childQuery = '';

		if (curType.getFields) {
			const crossReferenceKey = `${curParentName}To${curName}Key`;
			if (crossReferenceKeyList.indexOf(crossReferenceKey) !== -1 || curDepth > depthLimit) return '';
			crossReferenceKeyList.push(crossReferenceKey);
			const childKeys = Object.keys(curType.getFields());
			childQuery = childKeys
				.map(cur => generateQuery(cur, curType, curName, argumentsDict, duplicateArgCounts,
					crossReferenceKeyList, curDepth + 1, path.concat(curName)).queryStr)
				.filter(cur => cur)
				.join('\n');
		}

		if (!(curType.getFields && !childQuery)) {
			queryStr = `${'    '.repeat(curDepth)}${field.name}`;
			if (field.args.length > 0) {
				const dict = dedupe(field, duplicateArgCounts, argumentsDict, path);
				Object.assign(argumentsDict, dict);
				queryStr += `(${getArgsToVarsStr(dict)})`;
			}
			if (childQuery) {
				queryStr += `{\n${childQuery}\n${'    '.repeat(curDepth)}}`;
			}
		}

		/* Union types */
		if (curType.astNode && curType.astNode.kind === 'UnionTypeDefinition') {
			const types = curType.getTypes();
			if (types && types.length) {
				const indent = `${'    '.repeat(curDepth)}`;
				const fragIndent = `${'    '.repeat(curDepth + 1)}`;
				queryStr += '{\n';

				for (let i = 0, len = types.length; i < len; i++) {
					const valueTypeName = types[i];
					const valueType = schema.getType(valueTypeName);
					const unionChildQuery = Object.keys(valueType.getFields())
						.map(cur => generateQuery(cur, valueType, curName, argumentsDict, duplicateArgCounts,
							crossReferenceKeyList, curDepth + 2, path.concat(curName)).queryStr)
						.filter(cur => cur)
						.join('\n');
					queryStr += `${fragIndent}... on ${valueTypeName} {\n${unionChildQuery}\n${fragIndent}}\n`;
				}
				queryStr += `${indent}}`;
			}
		}
		return { queryStr, argumentsDict };
	};

	/**
	 * Generate the query for the specified field
	 * @param obj one of the root objects(Query, Mutation, Subscription)
	 * @param description description of the current object
	 */
	const addToResult = (obj, description) => {
		let field;
		switch (description) {
			case 'Mutation':
				field = 'mutations';
				break;
			case 'Query':
				field = 'queries';
				break;
			case 'Subscription':
				field = 'subscriptions';
				break;
			default:
				console.log('[gqlg warning]:', 'description is required');
		}
		result[field] = {};
		Object.keys(obj).forEach((type) => {
			result[field][type] = generateWrappedQuery({ type, description });
		});
	};

	function generateWrappedQuery({ type, description }) {
		const queryResult = generateQuery(type, description);
		const varsToTypesStr = getVarsToTypesStr(queryResult.argumentsDict);
		const query = queryResult.queryStr;
		return `${description.toLowerCase()} ${type}${varsToTypesStr ? `(${varsToTypesStr})` : ''}{\n${query}\n}`;
	}

	if (schema.getMutationType()) {
		addToResult(schema.getMutationType().getFields(), 'Mutation');
	} else {
		console.log('[gqlg warning]:', 'No mutation type found in your schema');
	}

	if (schema.getQueryType()) {
		addToResult(schema.getQueryType().getFields(), 'Query');
	} else {
		console.log('[gqlg warning]:', 'No query type found in your schema');
	}

	if (schema.getSubscriptionType()) {
		addToResult(schema.getSubscriptionType().getFields(), 'Subscription');
	} else {
		console.log('[gqlg warning]:', 'No subscription type found in your schema');
	}

	return result;
};