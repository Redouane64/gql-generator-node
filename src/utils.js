/* DOCUMENT INFORMATION
	- Author:   Dominik Maszczyk
	- Email:    Skitionek@gmail.com
	- Created:  2019-06-06
*/

import { MODULE_NAME } from "./constants";

/* istanbul ignore next */
export const moduleConsole = {
	log: (...args) =>
		console.log(`[${MODULE_NAME} log]:`, ...args),
	warn: (...args) =>
		console.log(`[${MODULE_NAME} warning]:`, ...args),
	error: (...args) =>
		console.log(`[${MODULE_NAME} error]:`, ...args)
};

/**
 * Compile arguments dictionary for a field
 * @param field current field object
 * @param duplicateArgCounts map for deduping argument name collisions
 * @param allArgsDict dictionary of all arguments
 * @param path to current field
 */
export const getFieldArgsDict = (
	field,
	duplicateArgCounts,
	allArgsDict = {},
	path
) =>
	field.args.reduce((o, arg) => {
		const arg_name = `${path.join('_')}_${field.name}_${arg.name}`;
		/* istanbul ignore next */
		if (arg_name in duplicateArgCounts) {
			moduleConsole.warn(`
			I cannot find the case for these duplicates anymore,
			please let me know if you are seeing this message.
			`);
			const index = duplicateArgCounts[arg_name] + 1;
			duplicateArgCounts[arg_name] = index;
			o[`${arg_name}${index}`] = arg;
		} else /* istanbul ignore next */ if (allArgsDict[arg_name]) {
			moduleConsole.warn(`
			I cannot find the case for these duplicates anymore,
			please let me know if you are seeing this message.
			`);
			duplicateArgCounts[arg_name] = 1;
			o[arg_name] = arg;
		} else if (!path.length) {
			o[arg.name] = arg;
		} else {
			o[arg_name] = arg;
		}
		return o;
	}, {});

/**
 * Generate variables string
 * @param dict dictionary of arguments
 */
export const getArgsToVarsStr = (dict, generateValues, requiredOnly) =>
	Object.entries(dict)
		.map(([fieldName, fieldInfo]) => 
			generateValues ? mapGeneratedArgsToVars(fieldName, fieldInfo, requiredOnly) :
			`${fieldInfo.name}: $${fieldName}`)
		.join(', ');

function mapGeneratedArgsToVars(fieldName, fieldTypeInfo, ignoreNonRequired) {
	const value = generateArgumentsValues(fieldTypeInfo, fieldName, ignoreNonRequired);
	return `${fieldTypeInfo.name}: ${value}`;
}

function generateArgumentsValues(parentFieldTypeInfo, parentFieldName = null, requiredOnly = false) {
	
	let value = "";

	const createArgumentValueRecursively = 
		(fieldTypeInfo, 
		fieldName = null, 
		_requiredOnly = false) => {

		let type = fieldTypeInfo.type;
		while (type.ofType) type = type.ofType;
		let _value = "";

		if (type.astNode && type.astNode.kind === 'InputObjectTypeDefinition') {
			_value += "{ ";
			_value += Object.entries(type.getFields()).map(([field, typeInfo]) => {
				return `${field}: ${createArgumentValueRecursively(typeInfo, field, _requiredOnly)}`;
			}).join(",");
			_value += " }";
		} else if (type.astNode && type.astNode.kind === 'EnumTypeDefinition') {
			if(_requiredOnly) {
				if (!fieldTypeInfo.type.toString().endsWith('!')) {
					return;
				}
			}
			_value += type.astNode.values[0].name.value;
		} else {
			if(_requiredOnly) {
				if (!fieldTypeInfo.type.toString().endsWith('!')) {
					return;
				}
			}

			const _v = generateScalarValue(fieldTypeInfo);
			console.log(_v);
			
			_value += "BOOOOOOM!";
		}

		return _value;
	}

	if (requiredOnly && !parentFieldTypeInfo.type.toString().endsWith('!')) {
		return null;
	}

	value += createArgumentValueRecursively(parentFieldTypeInfo, parentFieldTypeInfo.name, requiredOnly);
	return value;
}

function generateScalarValue(type) {
	let _type = type.astNode;
	let isRequired = false;
	while(_type.type) {
		if(_type.kind === 'NonNullType') {
			isRequired = true;
		}
		_type = _type.type;
	}

	if(isRequired) {
		return "BOOOOM!";
	}
}

/**
 * Generate types string
 * @param dict dictionary of arguments
 */
export const getVarsToTypesStr = dict =>
	Object.entries(dict)
		.map(([varName, arg]) => `$${varName}: ${arg.type}`)
		.join(', ');

