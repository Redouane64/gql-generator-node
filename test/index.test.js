require('graphql-import-node');

const typeDefs = require('./schemas/sampleTypeDef.graphql');
const typeDefsWithoutMutation = require('./schemas/empty.graphql');

const makeExecutableSchema = require('graphql-tools').makeExecutableSchema;
require('should');

const schema = makeExecutableSchema({ typeDefs });
const schemaWithoutMutation = makeExecutableSchema({ typeDefs: typeDefsWithoutMutation });

const schemaTypeDefs = require("./schemas/pineapple.graphql");
const testSchema = makeExecutableSchema({ typeDefs: schemaTypeDefs });

import { generateAll, generateQuery } from "../src";

it('validate generated queries', async () => {
	generateAll(schema, undefined, ({ args }) => {
		const o = {};
		(args || []).forEach(arg => {
			o[arg.name] = arg;
		});
		return o;
	}).mutations.signin.indexOf('signin').should.not.equal(-1)
});

it('limt depth', async () =>
	generateAll(schema, false, false, {}, 1).mutations.signup.indexOf('createdAt').should.equal(-1)
);

it('check field generator', async () =>
	expect(
		generateQuery({
			field: schema
				.getQueryType()
				.getFields().user
		})
	).toMatchSnapshot()
);

it('check field generator with skeleton', async () =>
	expect(
		generateQuery({
			field: schema
				.getQueryType()
				.getFields().user,
			skeleton: {
				email:
					true
			}
		})
	).toMatchSnapshot()
);

it('check field generator with skeleton - nested types', async () =>
	expect(
		generateQuery({
			field: schema
				.getQueryType()
				.getFields().user,
			skeleton: {
				email: true,
				context: {
					domain: true
				}
			}
		})
	).toMatchSnapshot()
);

it('check field generator with skeleton - unions', async () =>
	expect(
		generateQuery({
			field: schema
				.getQueryType()
				.getFields().user,
			skeleton: {
				email: true,
				details: {
					region: true
				}
			}
		})
	).toMatchSnapshot()
);

it('check field generator with skeleton - circular', async () =>
	expect(
		generateQuery({
			field: schema
				.getQueryType()
				.getFields().user,
			skeleton: {
				email: true,
				details: {
					user: {
						region: true
					},
					region: true
				}
			}
		})
	).toMatchSnapshot()
);

it('check field generator with skeleton - multiple fields with same query parameter name', async () =>
	expect(
		generateQuery({
			field: schema
				.getMutationType()
				.getFields().setConfig
		})
	).toMatchSnapshot()
);

it('check field generator with skeleton - multiple fields with same query parameter name', async () =>
	expect(
		generateQuery({
			field: schema
				.getMutationType()
				.getFields().setConfig,
			duplicateArgCounts: {
				setConfig_level_domain: "level_domain",
				setConfig_lastSeen_domain: "lastSeen_domain",
				setConfig_theme_domain: "theme_domain"
			}
		})
	).toMatchSnapshot()
);

it('check field generator for non-empty array ([]!)', async () =>
	expect(
		generateQuery({
			field: schema
				.getQueryType()
				.getFields().users,
			skeleton: {
				email:
					true
			}
		})
	).toMatchSnapshot()
);

it('check warnings for no mutations, query, subscription in schema', async () =>
	expect(
		generateAll(schemaWithoutMutation)
	).toMatchSnapshot()
);

it("Generate mock values", () => {

	const result = generateAll(testSchema, true, true)

	console.log(result);
})