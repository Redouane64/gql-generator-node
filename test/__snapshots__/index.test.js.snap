// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check field generator 1`] = `
"query user($user_context_user_details_region_language: String, $user_details_region_language: String, $id: Int!){
    user(id: $id){
        id
        username
        email
        createdAt
        context{
            user{
                id
                username
                email
                createdAt
                details{
                    ... on Guest {
                        region(language: $user_context_user_details_region_language)
                    }
                    ... on Member {
                        address
                    }
                }
            }
            domain
        }
        details{
            ... on Guest {
                region(language: $user_details_region_language)
            }
            ... on Member {
                address
            }
        }
    }
}"
`;

exports[`check field generator with skeleton 1`] = `
"query user($id: Int!){
    user(id: $id){
        email
    }
}"
`;
